server/supernova/audio_backend/portaudio_backend.hpp

```diff
@@ -121,15 +101,43 @@ class portaudio_backend : public detail::audio_backend_base<sample_type, float,
         }
     }
 
-    bool open_stream(std::string const& input_device, unsigned int inchans, std::string const& output_device,
-                     unsigned int outchans, unsigned int samplerate, unsigned int pa_blocksize, int h_blocksize) {
+    bool open_stream(std::string& input_device, unsigned int inchans, std::string& output_device, unsigned int outchans,
+                     unsigned int samplerate, unsigned int pa_blocksize, int h_blocksize) {
         int input_device_index, output_device_index;
-        if (!match_device(input_device, input_device_index) || !match_device(output_device, output_device_index))
+
+        input_device_index = GetPaDeviceFromName(input_device.c_str(), true);
+        output_device_index = GetPaDeviceFromName(output_device.c_str(), false);
+
+        // check device selection, select defaults if needed
+        fprintf(stdout, "\n");
+        PaError checked =
+            TryGetDefaultPaDevices(&input_device_index, &output_device_index, inchans, outchans, samplerate);
+        fprintf(stdout, "\n");
+
+        // if we got an error from CheckPaDevices, stop here
+        if (checked != paNoError) {
+            report_error(checked);
             return false;
+        }
+
+        // set device names
+        if (inchans)
+            input_device.assign(GetPaDeviceName(input_device_index));
+        if (outchans)
+            output_device.assign(GetPaDeviceName(output_device_index));
 
         PaStreamParameters in_parameters, out_parameters;
         PaTime suggestedLatencyIn, suggestedLatencyOut;
 
+        if (!samplerate) {
+            if (inchans)
+                samplerate = Pa_GetDeviceInfo(input_device_index)->defaultSampleRate;
+            if (outchans)
+                samplerate = Pa_GetDeviceInfo(output_device_index)->defaultSampleRate;
+            if (!samplerate)
+                samplerate = 44100;
+        }
```

Comment:
I don't think this is logically the same. In your suggestion, if both `inchans` and `outchans` are non-zero, sample rate from the input will be used; here the idea was to use output sample rate in that case (although this choice can be further discussed... also if the sample rate is not the same for both devices at this point, I'm not sure if opening the stream will succeed, but it might - you never know with PA, it depends on underlying API etc).

The final `if (!samplerate)` is not supposed to catch a situation where neither `inchans` nor `outchans` are non-zero, but rather if all previous calls to `PaDeviceInfo->defaultSampleRate` failed and sample rate is still not set.

Should I rewrite conditionals as `inchans != 0` and `samplerate == 0`?

Reply:
thanks for catching that; the inner set of conditionals can be rewritten though. let's save it for the next review round.

> Should I rewrite conditionals as inchans != 0 and samplerate == 0?

you can rewrite it here only if you feel like it, in the future that's the form to prefer though

Reply:
okay, now i understand what you mean, thank you for the explanation. this would do the same thing, i think, and is a little easier to understand:

```
if (outchans)
   ...
else if (inchans)
   ...
// Use 44100 as a last resort
if (samplerate == 0)
   samplerate = 44100;
```
personal style thing, but if there are a series of `if`s in a row, i like to separate them with a blank line, or comment:

```
if (a)
  something;

if (b)
  something;
if (c) // it's easy to mistake this for `else if`
  something;
```